File: ios/Classes/search/model/UnifiedDriveRouteResult.m
//
// Created by Yohom Bao on 2018-12-12.
//

#import "UnifiedDriveRouteResult.h"
#import "AMapSearchKit.h"
#import "NSArray+Rx.h"


@implementation UnifiedDriveRouteResult {

}
- (instancetype)initWithAMapRouteSearchResponse:(AMapRouteSearchResponse *)response {
    if ([self init]) {
        _startPos = response.route.origin;
        _targetPos = response.route.destination;
        _taxiCost = response.route.taxiCost;
        _paths = [response.route.paths map:^(AMapPath *path) {
            return [[UnifiedDrivePathResult alloc] initWithAMapPath:path];
        }];
    }
    return self;
}

@end


@implementation UnifiedDrivePathResult {
}

- (instancetype)initWithAMapPath:(AMapPath *)path {
    if ([self init]) {
        _strategy = path.strategy;
        _tolls = path.tolls;
        _tollDistance = path.tollDistance;
        _totalTrafficlights = path.totalTrafficLights;
        _steps = [path.steps map:^(AMapStep *step) {
            return [[UnifiedDriveStepResult alloc] initWithAMapStep:step];
        }];
        _restriction = path.restriction;
    }
    return self;
}

@end

@implementation UnifiedDriveStepResult {

}
- (instancetype)initWithAMapStep:(AMapStep *)step {
    if ([self init]) {
        _instruction = step.instruction;
        _orientation = step.orientation;
        _road = step.road;
        _distance = step.distance;
        _duration = step.duration;
        _polyline = step.polyline;
        _action = step.action;
        _assistantAction = step.assistantAction;
        _tolls = step.tolls;
        _tollDistance = step.tollDistance;
        _tollRoad = step.tollRoad;
        _routeSearchCityList = [step.cities map:^(AMapCity *city) {
            return [[UnifiedRouteSearchCityResult alloc] initWithAMapCity:city];
        }];
        _TMCs = [step.tmcs map:^(AMapTMC *tmc) {
            return [[UnifiedTMCResult alloc] initWithAMapTMC:tmc];
        }];
    }
    return self;
}

@end

@implementation UnifiedRouteSearchCityResult {

}
- (instancetype)initWithAMapCity:(AMapCity *)step {
    if ([self init]) {
        _districts = [step.districts map:^(AMapDistrict *district) {
            return [[UnifiedDistrictResult alloc] initWithAMapDistrict:district];
        }];
    }
    return self;
}

@end

@implementation UnifiedTMCResult {

}
- (instancetype)initWithAMapTMC:(AMapTMC *)tmc {
    if ([self init]) {
        _distance = tmc.distance;
        _status = tmc.status;
        _polyline = tmc.polyline;
    }
    return self;
}

@end

@implementation UnifiedDistrictResult {

}
- (instancetype)initWithAMapDistrict:(AMapDistrict *)district {
    if ([self init]) {
        _districtName = district.name;
        _districtAdcode = district.adcode;
    }
    return self;
}

@end

File: android/src/main/kotlin/me/yohom/amapbase/search/model/UnifiedDriveRouteResult.kt
package me.yohom.amapbase.search.model

import com.amap.api.maps.model.LatLng
import com.amap.api.services.route.*
import me.yohom.amapbase.search.toLatLng

internal class UnifiedDriveRouteResult(driveRouteResult: DriveRouteResult) {
    /// 起始
    val startPos: LatLng = driveRouteResult.startPos.toLatLng()

    /// 终点
    val targetPos: LatLng = driveRouteResult.targetPos.toLatLng()

    /// 打的费用
    val taxiCost: Double = driveRouteResult.taxiCost.toDouble()

    /// 路段
    val paths: List<UnifiedDrivePath> = driveRouteResult.paths.map { UnifiedDrivePath(it) }
}


class UnifiedDrivePath(drivePath: DrivePath) {
    val strategy: String = drivePath.strategy
    val tolls: Double = drivePath.tolls.toDouble()
    val tollDistance: Double = drivePath.tollDistance.toDouble()
    val totalTrafficlights: Int = drivePath.totalTrafficlights
    val steps: List<UnifiedDriveStep> = drivePath.steps.map { UnifiedDriveStep(it) }
    val restriction: Int = drivePath.restriction
}

class UnifiedDriveStep(driveStep: DriveStep) {
    val instruction: String = driveStep.instruction
    val orientation: String = driveStep.orientation
    val road: String = driveStep.road
    val distance: Double = driveStep.distance.toDouble()
    val tolls: Double = driveStep.tolls.toDouble()
    val tollDistance: Double = driveStep.tollDistance.toDouble()
    val tollRoad: String = driveStep.tollRoad
    val duration: Double = driveStep.duration.toDouble()
    val polyline: List<LatLng> = driveStep.polyline.map { it.toLatLng() }
    val action: String = driveStep.action
    val assistantAction: String = driveStep.assistantAction
    val routeSearchCityList: List<UnifiedRouteSearchCity> = driveStep.routeSearchCityList.map { UnifiedRouteSearchCity(it) }
    val TMCs: List<UnifiedTMC> = driveStep.tmCs.map { UnifiedTMC(it) }
}

class UnifiedRouteSearchCity(routeSearchCity: RouteSearchCity) {
    val districts: List<UnifiedDistrict> = routeSearchCity.districts.map { UnifiedDistrict(it) }
}

class UnifiedTMC(tmc: TMC) {
    val distance: Int = tmc.distance
    val status: String = tmc.status
    val polyline: List<LatLng> = tmc.polyline.map { it.toLatLng() }
}

class UnifiedDistrict(district: District) {
    val districtName: String = district.districtName
    val districtAdcode: String = district.districtAdcode
}